core/include/vecmem/memory/unique_ptr.hpp

```diff
@@ -0,0 +1,462 @@
+          m_align(i.m_align),
+          m_elems(i.m_elems) {}
+
+    /**
+     * @brief Activate the deletion mechanism of the deleter.
+     *
+     * @param p The pointer to deallocate.
+     *
+     * @note The pointer argument to this function is in some ways obsolete,
+     * because we store the pointer as a member variable. However, having both
+     * makes it easier to cross-check the logic.
+     */
+    void operator()(pointer_t p) const {
+        assert(p == m_ptr);
+        assert(m_mr != nullptr);
+
+        /*
+         * If this ever happens, something has gone VERY wrong...
+         */
+        if (m_mr == nullptr) {
+            return;
+        }
+
+        /*
+         * The class exhibits different behaviour for array types and non-array
+         * types.
+         */
+        if constexpr (!std::is_array_v<T>) {
+            /*
+             * If we are deleting a non-array type, we can simply call the
+             * stored object's destructor.
+             */
+            p->~T();
+        } else if constexpr (std::is_array_v<T>) {
+            /*
+             * If we are deleting an array type instead, we need to iterate
+             * over each of our stored elements and destruct them.
+             */
+            for (std::size_t i = 0; i < m_elems; ++i) {
+                p[i].~storage_t();
+            }
+        }
+
+        /*
+         * Finally, we need to deallocate the memory. For this, we use the
+         * memory resource, and we optionally pass it an alignment value.
+         */
+        if (m_align > 0) {
+            m_mr->deallocate(p, m_size, m_align);
+        } else {
+            m_mr->deallocate(p, m_size);
+        }
+    };
+
+private:
+    /*
+     * This should really be a reference, but using a reference here makes the
+     * class non-default destructible. At least - without some messy use of
+     * default memory resources. Using a pointer here is not ideal, but it
+     * allows us to make use of empty unique pointers.
+     */
+    memory_resource* m_mr;
+    pointer_t m_ptr;
+    std::size_t m_size;
+    std::size_t m_align;
+    std::size_t m_elems;
+};
+
+/**
+ * @brief A deleter class for trivial allocations.
+ *
+ * This class provides all the necessary functionality to allow unique pointers
+ * to deallocate allocations of trivial types, as well as arrays of them.
+ *
+ * @tparam T The type to deallocate.
+ *
+ * @note This deleter class supports non-array types, array types of unknown
+ * bound, but not array types of known bound.
+ *
+ * @note This deleter class only supports types that are both trivially
+ * constructible and trivially destructible.
+ */
+template <
+    typename T,
+    std::enable_if_t<std::is_trivially_constructible_v<std::remove_extent_t<T>>,
+                     bool> = true,
+    std::enable_if_t<std::is_trivially_destructible_v<std::remove_extent_t<T>>,
+                     bool> = true>
+struct unique_alloc_deleter {
```

Comment:
Just like for `unique_obj_deleter`, I would prefer to make this class non-templated. After all, it only has a single implementation anyway.

As before, we can put the SFINAE logic (without actual SFINAE...) into a traits class instead.

Reply:
Well, we need to make sure that `std::unique_ptr` only accepts certain kinds of types for its different partial applications, and the only way I see to do that elegantly would be to add SFINAE mechanisms to the deleter objects; that way, the specialization failures bubble up to the level of `std::unique_ptr` nicely, and that allows us to avoid having to reimplement `std::unique_ptr` completely. Do you have any other ideas of how you could achieve this without templating?

Reply:
Okay, I see what you mean - I think. I'll see if this can be done using a type trait.

Reply:
I played a bit just now. This may not be the greatest way of doing it, but it seems to (sort of...) work:

```c++
#include <iostream>
#include <memory>
#include <type_traits>

class A {
public:
   A() = default;

   int value() const { return m_value; }
   void setValue(int value) { m_value = value; }

private:
   int m_value;
};

class B {
public:
   B() = default;

   int value() const { return m_value; }
   void setValue(int value) { m_value = value; }

private:
   int m_value = 42;
};

struct unique_alloc_deleter {
   void operator()(void* ptr) {
      std::cout << "Deleting: " << ptr << std::endl;
   }
};

template <typename T>
struct unique_alloc_deleter_t {
  struct chosen_type_is_not_trivially_constructible {};
  using type = std::conditional_t<std::is_trivially_constructible<T>::value, unique_alloc_deleter,
				  chosen_type_is_not_trivially_constructible>;
};

template <typename T>
using unique_alloc_ptr = std::unique_ptr<T, typename unique_alloc_deleter_t<T>::type>;

int main() {

   std::cout << "A is trivially constructible: " << std::is_trivially_constructible<A>::value << std::endl;
   std::cout << "B is trivially constructible: " << std::is_trivially_constructible<B>::value << std::endl;

   unique_alloc_ptr<A> a(reinterpret_cast<A*>(0x1234));
   unique_alloc_ptr<B> b(reinterpret_cast<B*>(0x2345));

   return 0;
}
```

This produces the following compilation error:

```
[bash][Legolas]:cpp > g++ -std=c++17 trivial_construct.cpp 
In file included from /usr/include/c++/9/memory:80,
                 from trivial_construct.cpp:3:
/usr/include/c++/9/bits/unique_ptr.h: In instantiation of ‘std::unique_ptr<_Tp, _Dp>::~unique_ptr() [with _Tp = B; _Dp = unique_alloc_deleter_t<B>::chosen_type_is_not_trivially_constructible]’:
trivial_construct.cpp:63:54:   required from here
/usr/include/c++/9/bits/unique_ptr.h:288:56: error: static assertion failed: unique_ptr's deleter must be invocable with a pointer
  288 |  static_assert(__is_invocable<deleter_type&, pointer>::value,
      |                                                        ^~~~~
/usr/include/c++/9/bits/unique_ptr.h:292:17: error: no match for call to ‘(std::unique_ptr<B, unique_alloc_deleter_t<B>::chosen_type_is_not_trivially_constructible>::deleter_type {aka unique_alloc_deleter_t<B>::chosen_type_is_not_trivially_constructible}) (std::remove_reference<B*&>::type)’
  292 |    get_deleter()(std::move(__ptr));
      |    ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
[bash][Legolas]:cpp >
```

What's the exact error message with SFINAE? I just wonder which one would be more understandable...

Reply:
:confused: Honestly, what is the SFINAE output supposed to look like here? Since even with a pretty pure setup I don't get super readable error messages.

```c++
#include <iostream>
#include <memory>
#include <type_traits>

class A {
public:
   A() = default;

   int value() const { return m_value; }
   void setValue(int value) { m_value = value; }

private:
   int m_value;
};

class B {
public:
   B() = default;

   int value() const { return m_value; }
   void setValue(int value) { m_value = value; }

private:
   int m_value = 42;
};

struct unique_alloc_deleter {
   void operator()(void* ptr) {
      std::cout << "Deleting: " << ptr << std::endl;
   }
};

template <typename T, std::enable_if_t<std::is_trivially_constructible<T>::value, bool> = true>
using unique_alloc_ptr = std::unique_ptr<T, unique_alloc_deleter>;

int main() {

   std::cout << "A is trivially constructible: " << std::is_trivially_constructible<A>::value << std::endl;
   std::cout << "B is trivially constructible: " << std::is_trivially_constructible<B>::value << std::endl;

   unique_alloc_ptr<A> a(reinterpret_cast<A*>(0x1234));
   unique_alloc_ptr<B> b;

   return 0;
}
```

With this I get:

```
[bash][Legolas]:cpp > g++ -std=c++17 trivial_construct.cpp 
trivial_construct.cpp: In function ‘int main()’:
trivial_construct.cpp:56:22: error: no type named ‘type’ in ‘struct std::enable_if<false, bool>’
   56 |    unique_alloc_ptr<B> b;
      |                      ^
trivial_construct.cpp:56:22: note: invalid template non-type parameter
[bash][Legolas]:cpp >
```

Since we only provide a single implementation for this type, SFINAE does not seem to kick in anyway. So my previous proposal with a "descriptively named struct" may be the most readable output that we can generate...

Reply:
SFINAE never really gives you readable messages, does it? That only happens when you start using `static_assert`, and of course `static_assert` does not really participate in SFINAE. Anyway, these classes have some of their inner logic depend on template types as well. We would need to find a way to encapsulate that in a type trait - or reduce their parameter space and explicitly specialize them - in order to make them work.

Reply:
I would've expected a message like "`unique_alloc_ptr<B>` type is unknown", or something similar. This is what I would've understood by your "the specialization failures bubble up to the level of `std::unique_ptr` nicely" comment. :confused:

I don't know what you mean about the inner logic depending on the template type... For this class (`unique_alloc_deleter`) there is no type dependence whatsoever. For `unique_obj_deleter` you would have to split it into two classes. (As I wrote in that comment.)

So I'm still very much on the side of implementing these deleters as non-template classes. I just really don't like over-using templates... :frowning: (Which is unfortunately very hard not to do in this project.)

Reply:
Okay, I've looked at this a little more and I can see this being feasible for trivial types, but I am still a little bit miffed about how we can call the destructors of types without templating. _Maybe_ this would be possible if the type has virtual destructors, but that seems like a dangerous assumption to make. Do you have any ideas on how to make this work?

Reply:
You got me with the destructor... Indeed, `unique_obj_deleter` does need the concrete type to call the destructor correctly. So that one does need to remain a template... :thinking:

But `unique_alloc_deleter` does not need to do that. For this class we can really just call `deallocate` on the memory resource, and that's it. And that we can very much do through a `void*` pointer. And `std::unique_ptr` absolutely allows the deleter to receive a `void*` pointer.

Reply:
Alright, I've removed the templates as far as I deem possible, and I've moved the implementation to a separate `.cpp` file. I hope this is what you intended. :smile: